guides/Postgres Query Optimization + Diagnostic Tooling.md

```diff
@@ -0,0 +1,68 @@
+ &nbsp; 
+ > For Example:
+ 
+ &nbsp;&nbsp; `` EXPLAIN ANALYZE SELECT * FROM customer; ``
+ 
+  > Execution Plan of this Query is: <br>
+&nbsp;&nbsp;![](/.gitbook/assets/images/Picture1.png)
+ 
+## Postgres Query optimization and performance tuning methods:
+In this section, you will learn a few optimization techniques along with examples and query execution plans.
+### Use of Indexes:
+An index serves as an additional access structure that enables the rapid location and retrieval of specific data. In the absence of an index, a query would need to search the entire table to find the desired information. By indexing columns commonly used in WHERE or JOIN clauses, PostgreSQL can swiftly locate the needed data.
+
+For instance, the following picture depicts the query execution plan without any indexes.
+
+&nbsp;&nbsp; ![](/.gitbook/assets/images/Picture2.png)
+
+<br>The execution plan displayed in this image depicts the impact of adding an index on the "address_id" column. The addition of the index has resulted in a noticeable improvement in both Planning Time and Execution Time.
+
+&nbsp;&nbsp; ![](/.gitbook/assets/images/Picture3.png)
+
+### Optimize SELECT Clause:
+Efficiency issues in queries can be addressed by using specific column names instead of the ‘*’ argument in SELECT statements. Additionally, it is important to execute the HAVING clause after filtering the data with the SELECT statement, as SELECT acts as a filter. To illustrate, replacing ‘*’ with Payment_id in Group by can improve the query performance. You can see the change in execution plans in both the queries. <br><br>
+ &nbsp;&nbsp;   ![](/.gitbook/assets/images/Picture4.png)   
+ <br>
+ 
+ &nbsp;&nbsp; ![](/.gitbook/assets/images/Picture5.png) <br>
+ The performance difference is not significant, but according to the sources reviewed in the literature study, modifying the use of * is expected to enhance performance.
+ 
+ ### Filtering Rows with Subqueries and Indexes:
+ Filtering data before performing complex operations like GROUP BY or JOIN can enhance query performance by minimizing the amount of data that needs to be processed. Utilizing subqueries or other techniques to filter data is often effective in reducing query execution time and improving database efficiency. <br><br>
+&nbsp;&nbsp;   ![](/.gitbook/assets/images/Picture6.png)   
+ <br>
+ 
+ &nbsp;&nbsp; ![](/.gitbook/assets/images/Picture7.png)
+ ### WHERE Clause:
+ To optimize the WHERE Clause, we can use OR instead of IN when there is no index on the filtering column. The IN statement matches a column value to a list of values, and technically, it should execute faster than OR. However, in some cases, using OR instead of IN may not improve performance significantly, but it is a way to test if the query runs faster. <br><br>
```

Comment:
```suggestion
 To optimize the WHERE Clause, we can use OR instead of IN when there is no index on the filtering column. The IN statement matches a column value to a list of values, and technically, it should execute faster than OR. However, in some cases, using OR instead of IN may not improve performance significantly, but it is a way to test if the query can possibly run faster. <br><br>
```